refactor(cmd-socketio-server): remove code duplication

outSH opened the pull request:
- Moved verifyValidatorJwt to `cactus-common` to keep encrypt/decrypt logic
  of socketio-based validators in one, common location.
- Move config-reading and signMessageJwt helper functions from validators to cmd-socketio-server
  to remove code duplication.
  Refactor validators to use these common instead of own implementation.
- Remove ValidatorAuthentication.ts that is not used anymore
  (not part of public interface, it was copied by validators during before couple commits ago).
- Create jwt-message-authentication.test.ts from old, similar one in cactus-api-client.
- Updated readme with instructions of how to start asset-trade and electricity-trade samples
  without docker-compose (to be used during development).
  Added helper script for patching the config.

Signed-off-by: Michal Bajer <michal.bajer@fujitsu.com>

-----------------

This PR is for the last commit only - [refactor(cmd-socketio-server): remove code duplication](https://github.com/hyperledger/cactus/commit/200a7da23d68a38eb2c1af8869175e172c1d3b62). Please ignore previous one (dependencies) which will be removed before the merge.

Depends on #2033
Depends on #2129

jagpreetsinghsasan commented:
Thanks @outSH for your contribution.
It will take a while for me to review this, hope this ain't blocking anything super urgently.

outSH commented:
> Thanks @outSH for your contribution. It will take a while for me to review this, hope this ain't blocking anything super urgently.

@jagpreetsinghsasan Sure, there are bunch of depndencies awaiting to be merged first, no hurry ;) Just remember to look only at the last commit.

outSH commented:
Note to myself: Fix sample apps dockerfiles to use only npm packages

jagpreetsinghsasan reviewed (approved):
This looks all good to me.

petermetz commented on packages/cactus-common/package.json:
```diff
@@ -51,6 +51,7 @@
   },
   "dependencies": {
     "json-stable-stringify": "1.0.1",
+    "jsonwebtoken": "8.5.1",
```
@outSH Could you please swap this out with something that also works in the browser? (webpack builds)

outSH commented on packages/cactus-common/src/main/typescript/verifier/jwt-message-authentication.ts:
```diff
@@ -0,0 +1,92 @@
+ * @param targetData - Signed JWT message to be decoded.
+ * @returns Promise resolving to decoded jwt.JwtPayload.
+ */
+export function verifyValidatorMessageJwt(
```
@petermetz Yeah, this is unfortunate, I wanted to have whole sign logic in one common location, but didn't think about browser compatibility, sorry. I've removed this whole JWT section back to cmd-socketio-server and socketio-api-client like it was before. I hope to fully decomission them soon so no use of waisting more time on this than necessary.

This aplies to all 4 of your comments here

petermetz replied:
> @petermetz Yeah, this is unfortunate, I wanted to have whole sign logic in one common location, but didn't think about browser compatibility, sorry. I've removed this whole JWT section back to cmd-socketio-server and socketio-api-client like it was before. I hope to fully decomission them soon so no use of waisting more time on this than necessary.
> 
> This aplies to all 4 of your comments here

@outSH Makes sense, thank you for updating it! Just to re-iterate: I don't mind having a new package for it that we declare as "backend only". Exporting it from the cmd-api-server-socketio package is also fine by me but I imagine you might want to use these functions in some other packages that might not be able to import the API server due to circular dependencies.

outSH commented:
@petermetz Can you also publish cmd-sockio package from this branch to npm after #2129 is merged? It is needed by the connectors in this PR.

petermetz reviewed (approved):
@outSH LGTM, thank you!

petermetz reviewed (changes requested):
@outSH Please resolve the conflict and then pass it back for review!

outSH commented:
@petermetz I've rebased to the current main branch status, went through example apps startup (both docker and dockerless scenarios), fixed all errors and connector dockerfiles. Additionally, I've removed the dependency on Indy-SDK for main cactus build since it was tricky to install and sometimes their repo was not available. Updated the ci.sh and readme files.

For better understanding of changes I've done since review, you can check this commit: https://github.com/outSH/cactus/commit/de5dfef419a7482f65ca553b52e2aedf9dc731bf
(it was squashed in this PR).

outSH commented:
@petermetz OK, the tests are fixed, please merge when possible.

Changes since your last review: https://github.com/outSH/cactus/commit/083d91bd58caf2dc6b56395dd850bbae5f9c8bed

petermetz commented:
> @petermetz Can you also publish cmd-sockio package from this branch to npm after #2129 is merged? It is needed by the connectors in this PR.
> 
> // Note to myself: Test the dockerized sample apps manually after cmd-socketio package is published

@outSH Yes, sorry for not responding earlier, we now have the packages published to npm for v1.1.0 and v1.1.1 and v1.1.2 as well so you can use them with the new package names!

petermetz commented:
> @petermetz OK, the tests are fixed, please merge when possible.
> 
> Changes since your last review: [outSH@083d91b](https://github.com/outSH/cactus/commit/083d91bd58caf2dc6b56395dd850bbae5f9c8bed)

@outSH Thank you that commit link and the squash+rebase helped a lot with the re-review! LGTM

> Additionally, I've removed the dependency on Indy-SDK for main cactus build since it was tricky to install and sometimes their repo was not available. Updated the ci.sh and readme files.

Extra thanks for this because it was causing problems for people in ways that I could not reliably reproduce so it's much better that it is not an optional part of the build. Thank you!
